q2_moshpit/kraken2/classification.py

```diff
@@ -79,25 +81,149 @@ def _classify_kraken(
+                    bracken_db=str(bracken_db),
+                    kraken2_report_dir=report_dir,
+                    tmp_dir=tmpdir, threshold=threshold,
+                    read_len=read_len, level=level
+                )
+                bracken_tables.append(bracken_table)
+        except subprocess.CalledProcessError as e:
+            raise Exception(
+                "An error was encountered while running Bracken, "
+                f"(return code {e.returncode}), please inspect "
+                "stdout and stderr to learn more."
+            )
+
+    bracken_table = pd.concat(bracken_tables).reset_index()
+    bracken_table = bracken_table.pivot(
+        index="sample_id", columns="name", values="new_est_reads"
+    )
+
+    return bracken_table
+
+
+def _get_available_lens(bracken_db: BrackenDBDirectoryFormat) -> List[int]:
+    pattern = r'.+database(\d{2,})mers\.kmer_distrib$'
+    lengths = []
+    for db in bracken_db.path.iterdir():
+        lengths.extend(re.findall(pattern, str(db)))
+    return sorted([int(x) for x in lengths])
+
+
+def _is_reads(seqs):
+    return isinstance(
+            seqs, (SingleLanePerSamplePairedEndFastqDirFmt,
+                   SingleLanePerSampleSingleEndFastqDirFmt)
+    )
+
+
 def classify_kraken(
     seqs: Union[
         SingleLanePerSamplePairedEndFastqDirFmt,
         SingleLanePerSampleSingleEndFastqDirFmt,
         MultiMAGSequencesDirFmt,
     ],
-    db: Kraken2DBDirectoryFormat,
+    kraken2_db: Kraken2DBDirectoryFormat,
+    bracken_db: BrackenDBDirectoryFormat,
     threads: int = 1,
     confidence: float = 0.0,
     minimum_base_quality: int = 0,
     memory_mapping: bool = False,
     minimum_hit_groups: int = 2,
     quick: bool = False,
-) -> (Kraken2ReportDirectoryFormat, Kraken2OutputDirectoryFormat):
-    kwargs = {k: v for k, v in locals().items() if k not in ["seqs", "db"]}
+    threshold: int = 0,
+    read_len: int = 100,
+    level: str = 'S'
+) -> (
+        Kraken2ReportDirectoryFormat,
+        Kraken2OutputDirectoryFormat,
+        pd.DataFrame
+):
+    kraken2_args = [
+        "threads", "confidence", "minimum_base_quality",
+        "memory_mapping", "minimum_hit_groups", "quick"
+    ]
+    kwargs = {k: v for k, v in locals().items() if k in kraken2_args}
     common_args = _process_common_input_params(
         processing_func=_process_kraken2_arg, params=kwargs
     )
-    common_args.extend(["--db", str(db.path)])
+    common_args.extend(["--db", str(kraken2_db.path)])
     manifest: pd.DataFrame = seqs.manifest.view(pd.DataFrame)
 
-    return _classify_kraken(manifest, common_args)
+    if _is_reads(seqs):
+        # Check if read length is available in the Bracken DB
+        available_lens = _get_available_lens(bracken_db)
+        if read_len not in available_lens:
+            raise ValueError(
+                f"Provided read length ({read_len}) is not available in the "
+                f"Bracken DB. The available values are: "
```

Comment:
Is this something we could check for while validating input on import?

Reply:
I don't think we can create two artifacts: when building a Bracken DB you need to specify the read length. The databases _fetched_ by my other action contain several read lengths available and when users _build_ their own DBs  they can specify a list of lengths to build the DB with. I don't think we can do any validation on that either, as those DBs can contain an arbitrary number of read lengths (see my other [PR](https://github.com/bokulich-lab/q2-moshpit/pull/34))...